airbyte-oauth/src/main/java/io/airbyte/oauth/BaseOAuthFlow.java

```diff
@@ -133,57 +117,35 @@ private String getConsentUrl(UUID definitionId, String clientId, String redirect
                                                       Map<String, Object> queryParams,
                                                       String redirectUrl)
       throws IOException, ConfigNotFoundException {
-    if (queryParams.containsKey("code")) {
-      final String code = (String) queryParams.get("code");
-      final JsonNode oAuthParamConfig = getDestinationOAuthParamConfig(workspaceId, destinationDefinitionId);
-      final String clientId = getClientIdUnsafe(oAuthParamConfig);
-      final String clientSecret = getClientSecretUnsafe(oAuthParamConfig);
-      return completeOAuthFlow(clientId, clientSecret, code, redirectUrl);
-    } else {
-      throw new IOException("Undefined 'code' from consent redirected url.");
-    }
+    final JsonNode oAuthParamConfig = getDestinationOAuthParamConfig(workspaceId, destinationDefinitionId);
+    return completeOAuthFlow(
+        getClientIdUnsafe(oAuthParamConfig),
+        getClientSecretUnsafe(oAuthParamConfig),
+        extractCodeParameter(queryParams),
+        redirectUrl);
   }
 
-  protected Map<String, Object> completeOAuthFlow(String clientId, String clientSecret, String code, String redirectUrl) throws IOException {
-    final ImmutableMap<String, String> body = new Builder<String, String>()
-        .put("client_id", clientId)
-        .put("client_secret", clientSecret)
-        .put("code", code)
-        .put("grant_type", "authorization_code")
-        .put("redirect_uri", redirectUrl)
-        .build();
+  private Map<String, Object> completeOAuthFlow(String clientId, String clientSecret, String authCode, String redirectUrl) throws IOException {
     final HttpRequest request = HttpRequest.newBuilder()
-        .POST(HttpRequest.BodyPublishers.ofString(toUrlEncodedString(body)))
-        .uri(URI.create(ACCESS_TOKEN_URL))
+        .POST(HttpRequest.BodyPublishers.ofString(toUrlEncodedString(getAccessTokenQueryParameters(clientId, clientSecret, authCode, redirectUrl))))
+        .uri(URI.create(getAccessTokenUrl()))
         .header("Content-Type", "application/x-www-form-urlencoded")
         .build();
-    final HttpResponse<String> response;
     try {
-      response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
-      var data = Jsons.deserialize(response.body());
-      if (data.has("refresh_token")) {
-        return Map.of("refresh_token", data.get("refresh_token").asText());
-      } else {
-        // TODO This means the response from Google did not have a refresh token and is probably a
-        // programming error
-        // handle this better
-        throw new IOException(String.format("Missing 'refresh_token' in query params from %s. Response: %s", ACCESS_TOKEN_URL, data));
-      }
+      final HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());;
```

Comment:
I think the tricky thing here will be to parse an error response into an easily understood message for the user i.e: #6019 in combination with #6160